pkg/controller/command/command_upgrade.go

```diff
@@ -16,40 +16,59 @@ import (
+func (r *ReconcileCommand) performUpgradeStep(command *contrail.Command, expectedDeployment *apps.Deployment, configMapName string) (*apps.Deployment, error) {
+	if command.Status.UpgradeState == "" || command.Status.UpgradeState == contrail.CommandNotUpgrading {
+		command.Status.UpgradeState = contrail.CommandShuttingDownBeforeUpgrade
 	}
-	switch commandCR.Status.UpgradeState {
+	d := &apps.Deployment{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{
+		Namespace: command.GetNamespace(),
+		Name:      command.Name + "-command-deployment",
+	}, d); err != nil {
+		return nil, err
+	}
+	switch command.Status.UpgradeState {
 	case contrail.CommandShuttingDownBeforeUpgrade:
-		currentDeployment.Spec.Template.Spec.Containers = oldDeploymentSpec.Template.Spec.Containers
-		currentDeployment.Spec.Template.Spec.InitContainers = oldDeploymentSpec.Template.Spec.InitContainers
-		currentDeployment.Spec.Replicas = int32ToPtr(0)
-		if currentDeployment.Status.Replicas == 0 {
-			commandCR.Status.UpgradeState = contrail.CommandUpgrading
+		d.Spec.Replicas = int32ToPtr(0)
+		if d.Status.Replicas == 0 {
+			command.Status.UpgradeState = contrail.CommandUpgrading
 		}
 	case contrail.CommandUpgrading:
-		currentDeployment.Spec.Replicas = int32ToPtr(0)
-		oldImage := getContainerImage(oldDeploymentSpec.Template.Spec.Containers, "api")
-		newImage := getImage(commandCR.Spec.ServiceConfiguration.Containers, "api")
-		if err := r.reconcileDataMigrationJob(commandCR, oldImage, newImage, configMapName); err != nil {
-			return err
+		d.Spec.Replicas = int32ToPtr(0)
+		oldImage := getContainerImage(d.Spec.Template.Spec.Containers, "api")
+		newImage := getImage(command.Spec.ServiceConfiguration.Containers, "api")
+		if err := r.reconcileDataMigrationJob(command, oldImage, newImage, configMapName); err != nil {
+			return nil, err
+		}
+		if err := r.checkDataMigrationJobStatus(command); err != nil {
+			return nil, err
 		}
-		return r.checkDataMigrationJobStatus(commandCR)
 	case contrail.CommandStartingUpgradedDeployment:
-		if err := r.deleteMigrationJob(commandCR); err != nil {
-			return err
+		upgradedAgain, err := r.checkIfImageChangedAgain(command)
+		if err != nil {
+			return nil, err
+		}
+		if upgradedAgain {
+			command.Status.UpgradeState = contrail.CommandShuttingDownBeforeUpgrade
+			return d, nil
 		}
-		currentDeployment.Spec.Replicas = commandCR.Spec.CommonConfiguration.Replicas
-		expectedReplicas := ptrToInt32(currentDeployment.Spec.Replicas, 1)
-		if currentDeployment.Status.ReadyReplicas == expectedReplicas {
-			commandCR.Status.UpgradeState = contrail.CommandNotUpgrading
+		expectedDeployment.Spec.DeepCopyInto(&d.Spec)
+		d.Spec.Replicas = command.Spec.CommonConfiguration.Replicas
+		if err := r.prepareIntendedDeployment(d, command); err != nil {
+			return nil, err
+		}
+		expectedReplicas := ptrToInt32(d.Spec.Replicas, 1)
+		if d.Status.ReadyReplicas == expectedReplicas {
```

Comment:
Is it possible that we will have more than one ReadyReplica in that state?

Reply:
yes, we are in that state until whole deployment is ready